docs/proposals/diagnose-experience.md

```diff
@@ -0,0 +1,247 @@
+- **Content**: Verbose output includes lots of informational messages about the application's state, operations performed, configuration details, and more. They provide a broader view, which is helpful for tracing the overall execution of the tool.
+- **Level of Detail**: Very detailed, but usually focused on normal operations rather than errors or issues.
+
+### Debug Logs
+
+Debug logs focus on providing technical details for in-depth diagnosing and troubleshooting issues. It is intended for developers or technical users who need to understand the inner workings of the tool. Debug logs are detailed and technical, often including HTTP request and response from interactions between client and server, as well as code-specific information.
+
+- **Purpose**: Debug logs are specifically aimed at helping developers diagnose and fix issues within the application. They contain detailed technical information that is useful for troubleshooting problems.
+- **Target users**: Primarily intended for developers or technical users who are trying to understand the inner workings of the code and identify the root cause of issues.
+- **Content**: Debug logs focus on providing context needed to troubleshoot issues, like variable values, execution paths, error stack traces, and internal states of the application.
+- **Level of Detail**: Extremely detailed, providing insights into the application's internal workings and logic, often including low-level details that are essential for debugging.
+
+## Guiding Principles
+
+Here are the guiding principles to print out debug logs.
+
+### 1. **Timestamp Each Log Entry**
+- **Precise Timing:** Ensure each log entry has a precise timestamp to trace the sequence of events accurately.
+  - Example: `DEBUG: [2023-10-01T12:00:00Z] Starting metadata retrieval for repository oras-demo`
+
+### 2. **Capture API-Specific Details**
+- **API Requests:** Log detailed information about API requests made to the registry server, including the HTTP method, endpoint, headers, and body (excluding sensitive information).
+  - Example: `DEBUG: [HTTPRequest] POST /v2/oras-demo/blobs/uploads/ Headers: {Content-Length: 524288}`
+  
+- **API Responses:** Log details about the API responses received, including status codes, headers, and response body (excluding sensitive information).
+  - Example: `DEBUG: [HTTPResponse] Status: 201 Created, Headers: {Location: /v2/oras-demo/blobs/uploads/uuid}, Body: {}`
+
+### 3. **Log Before and After Critical Operations**
+- **Operation Logs:** Log before performing critical operations and after completing them, including success or failure status.
+  - Example: `DEBUG: Starting upload of layer 2 of 3 for repository oras-demo`
+  - Example: `DEBUG: Successfully uploaded layer 2 of 3 for repository oras-demo`
+
+- **State Logs:** Log important state information before and after key operations or decisions.
+  - Example: `DEBUG: Current retry attempt: 1, Max retries: 3`
```

Comment:
Do we also want to mention function names, important variable state, etc?

Reply:
ACK. Updated